staging/src/k8s.io/apiserver/pkg/server/options/etcd.go

```diff
@@ -223,12 +228,58 @@ func (s *EtcdOptions) Complete(storageObjectCountTracker flowcontrolrequest.Stor
 	}
 
 	if len(s.EncryptionProviderConfigFilepath) != 0 {
-		transformerOverrides, kmsPluginHealthzChecks, err := encryptionconfig.LoadEncryptionConfig(s.EncryptionProviderConfigFilepath, s.EncryptionProviderConfigAutomaticReload, stopCh)
+		ctx, closeTransformers := wait.ContextForChannel(stopCh)
+
+		encryptionConfiguration, err := encryptionconfig.LoadEncryptionConfig(s.EncryptionProviderConfigFilepath, s.EncryptionProviderConfigAutomaticReload, ctx.Done())
 		if err != nil {
+			// in case of error, we want to close partially initialized (if any) transformers
+			closeTransformers()
 			return err
 		}
-		s.transformerOverrides = transformerOverrides
-		s.kmsPluginHealthzChecks = kmsPluginHealthzChecks
+
+		s.kmsPluginHealthzChecks = encryptionConfiguration.HealthChecks
```

Comment:
it's confusing to set this when it's not going to be used in the reloading case

I think we should only have `s.kmsPluginHealthzChecks`, and set it conditional on whether we're reloading or not

```go
var err error
if s.EncryptionProviderConfigAutomaticReload {
  s.dynamicTransformers, err = encryptionconfig.NewReloadingDynamicTransformers(encryptionConfiguration, closeTransformers)
  if err != nil {
    return err
  }
  s.kmsPluginHealthzChecks = []healthz.HealthChecker{s.dynamicTransformers}
  err = addPostStartHook(...
    ...
} else {
  s.dynamicTransformers = encryptionconfig.NewDynamicTransformers(encryptionConfiguration.Transformers)
  s.kmsPluginHealthzChecks = encryptionConfiguration.HealthChecks
}
```

`NewReloadingDynamicTransformers` can ensure there's a single non-nil health check, etc

EtcdOptions#addEtcdHealthEndpoint can stay unchanged and just do `c.AddHealthChecks(s.kmsPluginHealthzChecks...)`